File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/user/UserDataRepository.java
package com.heaton.funnyvote.data.user;

import android.util.Log;

import com.heaton.funnyvote.data.RemoteServiceApi;
import com.heaton.funnyvote.database.User;
import com.heaton.funnyvote.retrofit.Server;

import java.io.IOException;

import io.reactivex.Observable;
import okhttp3.ResponseBody;
import retrofit2.Callback;

public class UserDataRepository implements UserDataSource {

    private static final String TAG = UserDataRepository.class.getSimpleName();
    private static UserDataRepository INSTANCE = null;

    private UserDataSource localUserDataSource;
    private UserDataSource remoteUserSource;

    public static UserDataRepository getInstance(UserDataSource localuserDataSource
            , UserDataSource remoteUserSource) {
        if (INSTANCE == null) {
            synchronized (UserDataRepository.class) {
                if (INSTANCE == null) {
                    INSTANCE = new UserDataRepository(localuserDataSource
                            , remoteUserSource);
                }
            }
        }
        return INSTANCE;
    }

    public static void destroyInstance() {
        INSTANCE = null;
    }

    public UserDataRepository(UserDataSource userDataSource, UserDataSource remoteUserSource) {
        this.localUserDataSource = userDataSource;
        this.remoteUserSource = remoteUserSource;
    }

    @Override
    public Observable<User> getUser(boolean forceUpdateUserCode) {
        final User user = localUserDataSource.getUser();
        if (user.getType() == User.TYPE_GUEST && (user.getUserCode() == null || user.getUserCode().isEmpty())) {
            final String guestName = "Guest" + (int) (Math.random() * 1000);
            Log.d(TAG, "Guest!" + user.getUserCode() + " name:" + guestName);
            return remoteUserSource.getGuestUserCode(guestName)
                    .flatMap(userCode -> {
                        user.setUserName(guestName);
                        user.setUserCode(userCode);
                        localUserDataSource.setUser(user);
                        return Observable.just(localUserDataSource.getUser());
                    });
        } else {
            if (forceUpdateUserCode) {
                return remoteUserSource.getUserInfo(user)
                        .flatMap(userDataQuery -> {
                            String userCode = "";
                            if (user.getType() == User.TYPE_GUEST) {
                                userCode = userDataQuery.guestCode;
                            } else {
                                userCode = userDataQuery.otp;
                            }
                            if (userCode != null) {
                                user.setUserCode(userCode);
                                localUserDataSource.setUser(user);
                                return Observable.just(localUserDataSource.getUser());
                            }
                            return Observable.error(new IOException("userCode is null"));
                        });
            } else {
                return Observable.just(user);
            }
        }
    }

    @Override
    public Observable<?> registerUser(String appId, User user, boolean mergeGuest) {
        String userType = "";
        switch (user.getType()) {
            case User.TYPE_FACEBOOK:
                userType = RemoteServiceApi.USER_TYPE_FACEBOOK;
                break;
            case User.TYPE_GOOGLE:
                userType = RemoteServiceApi.USER_TYPE_GOOGLE;
                break;
            case User.TYPE_TWITTER:
                userType = RemoteServiceApi.USER_TYPE_TWITTER;
                break;
            default:
                break;
        }
        if (!userType.isEmpty()) {
            final String guestCode = localUserDataSource.getUser().getUserCode();
            return remoteUserSource.getUserCode(userType, appId, user)
                    .flatMap(userCode -> {
                        if (mergeGuest) {
                            return remoteUserSource.linkGuestToLoginUser(userCode, guestCode)
                                    .map(responseBody -> {
                                        user.setUserCode(userCode);
                                        localUserDataSource.setUser(user);
                                        return userCode;
                                    });
                        } else {
                            user.setUserCode(userCode);
                            localUserDataSource.setUser(user);
                            return Observable.just(userCode);
                        }
                    });
        } else {
            Log.e(TAG, "registerUser onFailure: empty userType");
            return Observable.error(new Exception("registerUser onFailure"));
        }
    }

    @Override
    public void unregisterUser() {
        localUserDataSource.removeUser();
    }

    @Override
    public Observable<String> getUserCode(String userType, String appId, User user) {
        return remoteUserSource.getUserCode(userType, appId, user);
    }

    @Override
    public void linkGuestToLoginUser(String otp, String guest, Callback<ResponseBody> callback) {
        remoteUserSource.linkGuestToLoginUser(otp, guest, callback);
    }

    @Override
    public Observable<ResponseBody> linkGuestToLoginUser(String otp, String guest) {
        return remoteUserSource.linkGuestToLoginUser(otp, guest);
    }

    @Override
    public void changeUserName(Callback<ResponseBody> callback, String tokenType, String token, String name) {
        remoteUserSource.changeUserName(callback, tokenType, token, name);
    }

    @Override
    public Observable<ResponseBody> changeUserName(String tokenType, String token, String name) {
        return remoteUserSource.changeUserName(tokenType, token, name);
    }

    @Override
    public Observable<?> changeCurrentUserName(String name) {
        return getUser(false)
                .flatMap(user -> {
                    user.setUserName(name);
                    localUserDataSource.setUser(user);
                    return remoteUserSource.changeUserName(user.getTokenType(), user.getUserCode(), name);
                });
    }

    @Override
    public User getUser() {
        return localUserDataSource.getUser();
    }

    @Override
    public void setUser(User user) {
        localUserDataSource.setUser(user);
    }

    @Override
    public void removeUser() {
        localUserDataSource.removeUser();
    }

    @Override
    public Observable<String> getGuestUserCode(String name) {
        return remoteUserSource.getGuestUserCode(name);
    }

    @Override
    public void getUserInfo(Callback<Server.UserDataQuery> callback, User user) {
        remoteUserSource.getUserInfo(callback, user);
    }

    @Override
    public Observable<Server.UserDataQuery> getUserInfo(User user) {
        return remoteUserSource.getUserInfo(user);
    }

    @Override
    public void setGuestName(String guestName) {
        localUserDataSource.setGuestName(guestName);
    }
}

File: funnyvote/app/src/main/java/com/heaton/funnyvote/data/user/UserDataRepository.kt
package com.heaton.funnyvote.data.user

import android.util.Log
import com.heaton.funnyvote.data.RemoteServiceApi
import com.heaton.funnyvote.database.User
import com.heaton.funnyvote.retrofit.Server
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException

class UserDataRepository(
        private val localUserDataSource: UserDataSource,
        private val remoteUserSource: UserDataSource
) : UserDataSource {

    override var user: User
        get() = localUserDataSource.user
        set(user) {
            localUserDataSource.user = user
        }

    override fun getUser(callback: UserDataSource.GetUserCallback, forceUpdateUserCode: Boolean) {
        val user = localUserDataSource.user
        if (user.type == User.TYPE_GUEST && user.userCode.isEmpty()) {
            val guestName = "Guest" + (Math.random() * 1000).toInt()//Util.randomUserName(context);
            Log.d(TAG, "Guest!" + user.userCode + " name:" + guestName)
            remoteUserSource.getGuestUserCode(object : UserDataSource.GetUserCodeCallback {
                override fun onSuccess(userCode: String) {
                    user.userName = guestName
                    user.userCode = userCode
                    localUserDataSource.user = user
                    callback.onResponse(localUserDataSource.user)
                }

                override fun onFalure() {
                    callback.onFailure()
                }
            }, guestName)
        } else {
            if (forceUpdateUserCode) {
                remoteUserSource.getUserInfo(object : Callback<Server.UserDataQuery> {
                    override fun onResponse(call: Call<Server.UserDataQuery>, response: Response<Server.UserDataQuery>) {
                        if (response.isSuccessful) {
                            val userCode = if (user.type == User.TYPE_GUEST) {
                                response.body().guestCode
                            } else {
                                response.body().otp
                            }

                            if (userCode != null) {
                                user.userCode = userCode
                                localUserDataSource.user = user
                                callback.onResponse(localUserDataSource.user)
                            }
                        } else {
                            try {
                                val errorMessage = response.errorBody().string()
                                Log.e(TAG, "getUser onResponse false$errorMessage")
                            } catch (e: IOException) {
                                e.printStackTrace()
                            }

                            callback.onFailure()
                        }
                    }

                    override fun onFailure(call: Call<Server.UserDataQuery>, t: Throwable) {
                        callback.onFailure()
                    }
                }, user)
            } else {
                callback.onResponse(user)
            }
        }
    }

    override fun registerUser(appId: String, user: User, mergeGuest: Boolean, callback: UserDataSource.RegisterUserCallback) {
        var userType = ""
        when (user.type) {
            User.TYPE_FACEBOOK -> userType = RemoteServiceApi.USER_TYPE_FACEBOOK
            User.TYPE_GOOGLE -> userType = RemoteServiceApi.USER_TYPE_GOOGLE
            User.TYPE_TWITTER -> userType = RemoteServiceApi.USER_TYPE_TWITTER
        }
        if (!userType.isEmpty()) {
            val guestCode = localUserDataSource.user.userCode
            remoteUserSource.getUserCode(userType, appId, user, object : UserDataSource.GetUserCodeCallback {
                override fun onSuccess(userCode: String) {
                    if (mergeGuest) {
                        remoteUserSource.linkGuestToLoginUser(userCode, guestCode
                                , object : Callback<ResponseBody> {
                            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                                if (response.isSuccessful) {
                                    user.userCode = userCode
                                    localUserDataSource.user = user
                                    callback.onSuccess()
                                } else {
                                    callback.onFailure()
                                    try {
                                        Log.e(TAG, "registerUser" + response.errorBody().string())
                                    } catch (e: Exception) {
                                        e.printStackTrace()
                                    }

                                }
                            }

                            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                                callback.onFailure()
                                Log.e(TAG, "registerUser onFailure , error message:" + t.message)
                            }
                        })
                    } else {
                        user.userCode = userCode
                        localUserDataSource.user = user
                        callback.onSuccess()
                    }
                }

                override fun onFalure() {
                    callback.onFailure()
                    Log.e(TAG, "registerUser onFailure")
                }
            })

        } else {
            callback.onFailure()
            Log.e(TAG, "registerUser onFailure")
        }
    }

    override fun unregisterUser() {
        localUserDataSource.removeUser()
    }

    override fun getUserCode(userType: String, appId: String, user: User, callback: UserDataSource.GetUserCodeCallback) {
        remoteUserSource.getUserCode(userType, appId, user, callback)
    }

    override fun linkGuestToLoginUser(otp: String, guest: String, callback: Callback<ResponseBody>) {
        remoteUserSource.linkGuestToLoginUser(otp, guest, callback)
    }

    override fun changeUserName(callback: Callback<ResponseBody>, tokenType: String, token: String, name: String) {
        // only for remote
    }

    private fun changeUserName(user: User, newName: String, callback: UserDataSource.ChangeUserNameCallback) {
        remoteUserSource.changeUserName(object : Callback<ResponseBody> {
            override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
                if (response.isSuccessful) {
                    user.userName = newName
                    localUserDataSource.user = user
                    callback.onSuccess()
                } else {
                    try {
                        Log.d(TAG, "changeUserName response status:" + response.code() + " ,message:" + response.errorBody().string())
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                    callback.onFailure()
                }

            }

            override fun onFailure(call: Call<ResponseBody>, t: Throwable) {
                t.printStackTrace()
                callback.onFailure()
            }
        }, user.tokenType, user.userCode, newName)
    }

    override fun changeCurrentUserName(name: String, callback: UserDataSource.ChangeUserNameCallback) {
        getUser(object : UserDataSource.GetUserCallback {
            override fun onResponse(user: User) {
                changeUserName(user, name, callback)
            }

            override fun onFailure() {
                Log.w(TAG, "changeUserName Fail")
            }
        }, true)
    }

    override fun removeUser() {
        localUserDataSource.removeUser()
    }

    override fun getGuestUserCode(callback: UserDataSource.GetUserCodeCallback, name: String) {
        remoteUserSource.getGuestUserCode(callback, name)
    }

    override fun getUserInfo(callback: Callback<Server.UserDataQuery>, user: User) {
        remoteUserSource.getUserInfo(callback, user)
    }

    override fun setGuestName(guestName: String) {

    }

    companion object {

        @JvmField val TAG: String? = UserDataRepository::class.java.simpleName
        private var INSTANCE: UserDataRepository? = null

        @JvmStatic
        fun getInstance(localUserDataSource: UserDataSource
                        , remoteUserDataSource: UserDataSource): UserDataRepository {
            return INSTANCE ?: UserDataRepository(localUserDataSource, remoteUserDataSource)
                    .apply { INSTANCE = this }
        }

        @JvmStatic
        fun destroyInstance() {
            INSTANCE = null
        }
    }
}
